ouroboros-consensus/src/Ouroboros/Consensus/Util/CBOR.hs

```diff
@@ -162,43 +163,55 @@ readIncrementalOffsets hasFS@HasFS{..} decoder fp = withLiftST $ \liftST ->
-             go liftST h 0 [] Nothing fileSize
+             go liftST h 0 [] Nothing [] fileSize
   where
     go :: (forall x. ST s x -> m x)
        -> h
        -> Word64                  -- ^ Offset
        -> [(Word64, (Word64, a))] -- ^ Already deserialised (reverse order)
        -> Maybe ByteString        -- ^ Unconsumed bytes from last time
+       -> [ByteString]            -- ^ Chunks pushed for this item (rev order)
        -> Word64                  -- ^ Total file size
-       -> CBOR.IDecode s a
+       -> CBOR.IDecode s (LBS.ByteString -> a)
        -> m ([(Word64, (Word64, a))], Maybe ReadIncrementalErr)
-    go liftST h offset deserialised mbUnconsumed fileSize dec = case dec of
+    go liftST h offset deserialised mbUnconsumed bss fileSize dec = case dec of
       CBOR.Partial k -> do
         -- First use the unconsumed bytes from a previous read before read
         -- some more bytes from the file.
         bs   <- case mbUnconsumed of
           Just unconsumed -> return unconsumed
           Nothing         -> hGetSome h (fromIntegral defaultChunkSize)
         dec' <- liftST $ k (checkEmpty bs)
-        go liftST h offset deserialised Nothing fileSize dec'
-
-      CBOR.Done leftover size !a -> do
-        -- The bang on the @a@ here allows the used 'Decoder' to force its
-        -- computation. For example, the decoder might decode a whole block
-        -- and then (maybe through a use of 'fmap') just return its hash. If
-        -- we don't force the value it returned here, we're just putting a
-        -- thunk that references the whole block in the list instead of merely
-        -- the hash.
+        go liftST h offset deserialised Nothing (bs:bss) fileSize dec'
+
+      CBOR.Done leftover size a -> do
         let nextOffset    = offset + fromIntegral size
-            deserialised' = (offset, (fromIntegral size, a)) : deserialised
+            -- We've been keeping track of the bytes pushed into the decoder
+            -- for this item so far in bss. Now there's some trailing data to
+            -- remove and we can get the whole bytes used for this item. We
+            -- supply the bytes to the final decoded value. This is to support
+            -- annotating values with their original input bytes.
+            aBytes        = case bss of
+                []      -> LBS.empty
+                bs:bss' -> LBS.fromChunks (reverse (bs' : bss'))
+                  where
+                    bs' = BS.take (BS.length bs - BS.length leftover) bs
+            -- The bang on the @a'@ here allows the used 'Decoder' to force
+            -- its computation. For example, the decoder might decode a whole
+            -- block and then (maybe through a use of 'fmap') just return its
+            -- hash. If we don't force the value it returned here, we're just
+            -- putting a thunk that references the whole block in the list
+            -- instead of merely the hash.
+            !a'            = a aBytes
```

Comment:
I think this will be ok:
```
data Annotated b a = Annotated { unAnnotated :: !b, annotation :: !a }
  deriving (Eq, Show, Functor, Generic)
```
So it's going to force the contained value to WHNF.